File: structural/proxy/image_loading/python/image_proxy.py
from abc import ABC, abstractmethod
import time

# --- Subject Interface --- #

class Image(ABC):
    """Subject Interface: Declares the common interface for RealSubject and Proxy."""
    @abstractmethod
    def display(self) -> None:
        """Method to display the image."""
        pass

    @abstractmethod
    def get_filename(self) -> str:
        """Returns the filename associated with the image."""
        pass

# --- Real Subject --- #

class RealImage(Image):
    """RealSubject: Defines the real object that the proxy represents.
    Loading the image is simulated here.
    """
    def __init__(self, filename: str):
        self._filename = filename
        self._load_from_disk()

    def _load_from_disk(self) -> None:
        """Private helper method to simulate loading the image data."""
        print(f"Loading image: '{self._filename}' from disk... (Simulating delay)")
        # Simulate time-consuming operation
        time.sleep(1.5)
        print(f"Finished loading image: '{self._filename}'")

    def display(self) -> None:
        """Displays the image (after it has been loaded)."""
        print(f"Displaying image: '{self._filename}'")

    def get_filename(self) -> str:
        """Returns the filename."""
        return self._filename

# --- Proxy --- #

class ProxyImage(Image):
    """Proxy: Maintains a reference that lets the proxy access the real subject.
    Implements the same interface as the RealSubject.
    Controls access to the real subject and may be responsible for its creation (lazy loading).
    """
    def __init__(self, filename: str):
        self._filename = filename
        self._real_image: RealImage | None = None # Reference to RealImage, initially None
        print(f"ProxyImage created for: '{self._filename}' (Real image not loaded yet)")

    def display(self) -> None:
        """Handles the display request.
        Loads the RealImage only if it hasn't been loaded yet (Lazy Initialization).
        """
        if self._real_image is None:
            print(f"Proxy for '{self._filename}': Real image needs loading.")
            # Lazy initialization: Create the RealImage object only when needed
            self._real_image = RealImage(self._filename)
        else:
            print(f"Proxy for '{self._filename}': Real image already loaded.")

        # Delegate the display call to the RealImage object
        self._real_image.display()

    def get_filename(self) -> str:
        """Returns the filename without loading the real image."""
        return self._filename

    # Optional: Method to check if loaded without triggering load
    def is_loaded(self) -> bool:
        """Checks if the real image has been loaded."""
        return self._real_image is not None

File: structural/proxy/image_loading/typescript/image_proxy.ts
// Utility function for async delay (simulating I/O)
const delay = (ms: number) => new Promise(resolve => setTimeout(resolve, ms));

// --- Subject Interface --- //

export interface IImage {
    /**
     * Subject Interface: Declares the common interface for RealSubject and Proxy.
     */
    display(): Promise<void>; // Display might involve loading, so make it async
    getFilename(): string;
}

// --- Real Subject --- //

export class RealImage implements IImage {
    /**
     * RealSubject: Defines the real object that the proxy represents.
     * Loading the image is simulated here as an async operation.
     */
    private filename: string;

    // Private constructor ensures it's typically created via a factory or proxy
    private constructor(filename: string) {
        this.filename = filename;
        // Loading happens immediately in the 'create' method
    }

    // Static factory method to simulate async loading during creation
    public static async create(filename: string): Promise<RealImage> {
        const image = new RealImage(filename);
        await image.loadFromDisk();
        return image;
    }

    private async loadFromDisk(): Promise<void> {
        /**
         * Private helper method to simulate loading the image data asynchronously.
         */
        console.log(`Loading image: '${this.filename}' from disk... (Simulating async delay)`);
        await delay(1500); // Simulate time-consuming I/O
        console.log(`Finished loading image: '${this.filename}'`);
    }

    public async display(): Promise<void> {
        /**
         * Displays the image (after it has been loaded).
         */
        console.log(`Displaying image: '${this.filename}'`);
        // In a real scenario, this might render to a canvas, etc.
    }

    public getFilename(): string {
        /**
         * Returns the filename.
         */
        return this.filename;
    }
}

// --- Proxy --- //

export class ProxyImage implements IImage {
    /**
     * Proxy: Maintains a reference that lets the proxy access the real subject.
     * Implements the same interface as the RealSubject.
     * Controls access to the real subject and is responsible for its creation (lazy loading).
     */
    private filename: string;
    private realImage: RealImage | null = null; // Reference to RealImage, initially null

    constructor(filename: string) {
        this.filename = filename;
        console.log(`ProxyImage created for: '${this.filename}' (Real image not loaded yet)`);
    }

    public async display(): Promise<void> {
        /**
         * Handles the display request.
         * Loads the RealImage asynchronously only if it hasn't been loaded yet (Lazy Initialization).
         */
        if (this.realImage === null) {
            console.log(`Proxy for '${this.filename}': Real image needs loading.`);
            // Lazy initialization: Create and load the RealImage object only when needed
            this.realImage = await RealImage.create(this.filename);
        } else {
            console.log(`Proxy for '${this.filename}': Real image already loaded.`);
        }

        // Delegate the display call to the RealImage object
        // No need to await here as the display logic in RealImage itself is synchronous
        // If RealImage.display() were async, we would await here.
        await this.realImage.display();
    }

    public getFilename(): string {
        /**
         * Returns the filename without loading the real image.
         */
        return this.filename;
    }

    // Optional: Method to check if loaded without triggering load
    public isLoaded(): boolean {
        /**
         * Checks if the real image has been loaded.
         */
        return this.realImage !== null;
    }
}
